README.md
Calendar component for Ember.js.

---

© 2013 [SpotOn](https://spoton.it), shared under the [MIT License](http://www.opensource.org/licenses/MIT).

HISTORY.md
1.0.0 / 2013-10-11
==================

* Added daily view and view switching.
* Added convertible example demonstrating view switching.
* Renamed `Ember.Calendar.CalendarView` to `Ember.Calendar.ContainerView`; added new `Ember.Calendar.CalendarView` that wraps (optional) view switching buttons and calendar body.
* Renamed weekly view template: `ember-calendar` -> `ember-calendar-week`.
* Renamed keys related to weekly view in Ember.Calendar.CalendarController: `dayViewClass` -> `weekDayViewClass`, `headingDateViewClass` -> `weekHeadingDateViewClass`, `dates` -> `weekDates`, `days` -> `weekDays`.

build.js
require('shelljs/global')

cd(__dirname)

rm('-r', 'dist')
mkdir('dist')

var license = '/* ember-calendar (https://github.com/joinspoton/ember-calendar) | (c) 2013 SpotOn (https://spoton.it) | http://www.opensource.org/licenses/MIT */\n'

var handlebars = cat('src/calendar.html').split('</script>').slice(0, -1).map(function (source) {
	var name = source.match(/data-template-name=\'(.*)\'/)[1]
	
	source = source.replace(/<script.*/gm, '')
	source = JSON.stringify(source)
	
	return 'Ember.TEMPLATES["' + name + '"]=Ember.Handlebars.compile(' + source + ');\n'
}).join('')

;(license + cat('src/calendar.css')).to('dist/ember-calendar.css')
;(license + handlebars + cat('src/calendar.js')).to('dist/ember-calendar.js')

;(license + exec('yuicompressor dist/ember-calendar.css', { silent: true }).output).to('dist/ember-calendar.min.css')
;(license + exec('closure-compiler --language_in ECMASCRIPT5 --js dist/ember-calendar.js', { silent: true }).output).to('dist/ember-calendar.min.js')

javascripts/app.js
///////////////////////////////////////////////////////////////////////////////
// Application
///////////////////////////////////////////////////////////////////////////////
App = Ember.Application.create();

App.ApplicationController = Ember.Controller.extend({
    needs: ['convertibleCalendar', 'multitypeCalendar', 'ajaxCalendar']
});

App.ApplicationRoute = Ember.Route.extend({
    setupController: function () {
      this.controllerFor('ajaxCalendar').update();
    }
});

App.ApplicationView = Ember.View.extend({
    templateName: 'application'
});


///////////////////////////////////////////////////////////////////////////////
// Views
///////////////////////////////////////////////////////////////////////////////
App.MultitypeEventView = Ember.Calendar.EventView.extend({
    templateName: function () {
      return this.get('event.template') || 'ember-calendar-event';
    }.property('event.template')
    
  , classNameBindings: ['facebook', 'google', 'spoton']
  , spoton: function () {
      return this.get('event.type') === 0;
    }.property('event.type')
  , google: function () {
      return this.get('event.type') === 1;
    }.property('event.type')
  , facebook: function () {
      return this.get('event.type') === 2;
    }.property('event.type')
});


///////////////////////////////////////////////////////////////////////////////
// Controllers
///////////////////////////////////////////////////////////////////////////////
App.ConvertibleCalendarController = Ember.Calendar.CalendarController.extend({
    states: ['day', 'week']
  , initialState: 'week'
  , content: function () {
      var events = [];
      var date;
      var time;
      var duration;
      var i;
      
      for (i = 0; i < 45; i++) {
        date = Math.floor(Math.random() * 21) - 7;
        time = 1000 * 60 * 60 * 8 + 1000 * 60 * 30 * Math.floor(Math.random() * 24);
        duration = 1000 * 60 * 30 * (1 + Math.floor(Math.random() * 5));
        
        events.push({
            name: 'Event ' + events.length
          , start: moment().startOf('day').add('days', date - moment().day()).add('milliseconds', time)
          , end: moment().startOf('day').add('days', date - moment().day()).add('milliseconds', time + duration)
        });
      }
      
      return events;
    }.property()
});

App.AjaxCalendarController = Ember.Calendar.CalendarController.extend({
    content: []
  , update: function () {
      if (!this.get('week')) { return; }
      
      var self = this;
      $.getJSON('http://ember-calendar-ajax.herokuapp.com', { week: self.get('week').toDate() }, function (response) {
        if (!response) { return; }
        self.clear().pushObjects(response).notifyPropertyChange('content');
      });
    }.observes('week')
});

App.MultitypeCalendarController = Ember.Calendar.CalendarController.extend({
    content: function () {
      var events = [];
      var date;
      var time;
      var duration;
      var event;
      var type;

      for (var i = 0; i < 10; i++) {
        date = Math.floor(Math.random() * 7);
        time = 1000 * 60 * 60 * 8 + 1000 * 60 * 30 * Math.floor(Math.random() * 24);
        duration = 1000 * 60 * 30 * (3 + Math.floor(Math.random() * 5));
        
        event = {
            name: 'Event ' + events.length
          , start: moment().startOf('day').add('days', date - moment().day()).add('milliseconds', time)
          , end: moment().startOf('day').add('days', date - moment().day()).add('milliseconds', time + duration)
          , type: Math.floor(Math.random() * 3) // 0 = spoton, 1 = google, 2 = facebook
        };
        
        // add unique properties for spoton type 
        if (event.type === 0) {
          event.numGoing = Math.floor(Math.random() * 100);
          event.template = 'multitype-event-spoton';
        }
        
        events.push(event);
      }
      
      return events;
    }.property()
    
  , eventViewClass: 'App.MultitypeEventView'
});

javascripts/ember-calendar.js
/* ember-calendar (https://github.com/joinspoton/ember-calendar) | (c) 2013 SpotOn (https://spoton.it) | http://www.opensource.org/licenses/MIT */
Ember.TEMPLATES["ember-calendar"]=Ember.Handlebars.compile("\n  <div class='ember-calendar-head'>\n    <div class='ember-calendar-head-previous'><button {{action loadPreviousWeek target='controller'}}>&laquo;</button></div>\n    {{view Ember.Calendar.HeadingDatesView datesBinding='controller.dates'}}\n    <div class='ember-calendar-head-next'><button {{action loadNextWeek target='controller'}}>&raquo;</button></div>\n  </div>\n  <div class='ember-calendar-container'>\n    <div class='ember-calendar-body'>\n      {{view Ember.Calendar.HeadingTimesView timesBinding='controller.times'}}\n      {{view Ember.Calendar.DaysView daysBinding='controller.days'}}\n    </div>\n  </div>\n");
Ember.TEMPLATES["ember-calendar-head-date"]=Ember.Handlebars.compile("\n\n\n  {{view.dateString}}\n");Ember.TEMPLATES["ember-calendar-head-time"]=Ember.Handlebars.compile("\n\n\n  {{view.timeString}}\n");Ember.TEMPLATES["ember-calendar-event"]=Ember.Handlebars.compile("\n\n\n  <div class='ember-calendar-event-name'>{{view.nameString}}</div>\n  <div class='ember-calendar-event-time'>{{view.timeString}}</div>\n  <div class='ember-calendar-event-location'>{{view.locationString}}</div>\n");
Ember.Calendar=Ember.Namespace.create();
Ember.Calendar.CalendarController=Ember.ArrayController.extend({startOfWeek:0,startOfDay:8,headingDateFormat:"ddd MMM D",headingTimeFormat:"h a",headingTimeRangeStart:0,headingTimeRangeEnd:24,eventTimeFormat:"h:mm a",eventTimeSeparator:" - ",headingDateViewClass:"Ember.Calendar.HeadingDateView",headingTimeViewClass:"Ember.Calendar.HeadingTimeView",dayViewClass:"Ember.Calendar.DayView",eventViewClass:"Ember.Calendar.EventView",week:null,dates:function(){if(!this.get("week"))return[];var a=this.get("week").clone().subtract("days",
1),b=[],c;for(c=0;7>c;c++)b.push(a.add("days",1).clone());return b}.property("week"),times:function(){var a=[],b;for(b=this.get("headingTimeRangeStart");b<this.get("headingTimeRangeEnd");b++)a.push(36E5*b);return a}.property("headingTimeRangeStart","headingTimeRangeStart"),days:function(){if(!this.get("week"))return[];var a=[[],[],[],[],[],[],[]],b=this.get("dates"),c=this,f=b[0].clone(),k=b[6].clone().endOf("day");this.get("content").forEach(function(b){var e=moment(b.start).clone(),g=moment(b.end).clone(),
d,h;if(!(g<f||e>k))for(;g>e;)d={},Object.keys(b).forEach(function(a){"start"!==a&&"end"!==a&&(d[a]=b[a])}),d.start=e.clone(),d.end=e.clone().endOf("day"),d.end>g&&(d.end=g.clone()),h=d.start.clone().startOf("day").diff(c.get("week"),"days"),0<=h&&6>=h&&a[h].push(d),e.add("days",1).startOf("day")});return a}.property("content","dates"),loadPreviousWeek:function(){this.get("week").subtract("days",7);this.notifyPropertyChange("week")},loadNextWeek:function(){this.get("week").add("days",7);this.notifyPropertyChange("week")},
init:function(){this._super();this.set("week",moment(this.get("initialDate")).subtract("days",(moment(this.get("initialDate")).day()+7-this.get("startOfWeek"))%7).startOf("day"))}});
Ember.Calendar.CalendarView=Ember.View.extend({templateName:"ember-calendar",classNames:["ember-calendar"],didInsertElement:function(){var a=$("#"+this.get("elementId")+" > .ember-calendar-container"),b=$("#"+this.get("elementId")+" > .ember-calendar-container > .ember-calendar-body"),c=(this.get("controller.startOfDay")-this.get("controller.headingTimeRangeStart"))/(this.get("controller.headingTimeRangeEnd")-this.get("controller.headingTimeRangeStart"));a.scrollTop(c*b.height())}});
Ember.Calendar.HeadingDatesView=Ember.ContainerView.extend({classNames:["ember-calendar-head-dates"],updateChildViews:function(){var a=this;a.removeAllChildren();a.pushObjects(a.get("dates").map(function(b){return a.createChildView(Ember.get(a.get("parentView.controller.headingDateViewClass")),{date:b})}))}.observes("dates"),init:function(){this._super();this.updateChildViews()}});
Ember.Calendar.HeadingDateView=Ember.View.extend({templateName:"ember-calendar-head-date",classNames:["ember-calendar-head-date"],dateString:function(){return this.get("date").format(this.get("parentView.parentView.controller.headingDateFormat"))}.property("date")});
Ember.Calendar.HeadingTimesView=Ember.ContainerView.extend({classNames:["ember-calendar-head-times"],updateChildViews:function(){var a=this;a.removeAllChildren();a.pushObjects(a.get("times").map(function(b){return a.createChildView(Ember.get(a.get("parentView.controller.headingTimeViewClass")),{time:b})}))}.observes("times"),init:function(){this._super();this.updateChildViews()}});
Ember.Calendar.HeadingTimeView=Ember.View.extend({templateName:"ember-calendar-head-time",classNames:["ember-calendar-head-time"],timeString:function(){return moment().startOf("day").add("milliseconds",this.get("time")).format(this.get("parentView.parentView.controller.headingTimeFormat"))}.property("time")});
Ember.Calendar.DaysView=Ember.ContainerView.extend({classNames:["ember-calendar-days"],updateChildViews:function(){var a=this;a.removeAllChildren();a.pushObjects(a.get("days").map(function(b){return a.createChildView(Ember.get(a.get("parentView.controller.dayViewClass")),{events:b,parentView:a.get("parentView")})}))}.observes("days"),init:function(){this._super();this.updateChildViews()}});
Ember.Calendar.DayView=Ember.ContainerView.extend({classNames:["ember-calendar-day"],updateChildViews:function(){var a=this;a.removeAllChildren();a.pushObjects(a.get("events").map(function(b){return a.createChildView(Ember.get(a.get("parentView.controller.eventViewClass")),{event:b,parentView:a.get("parentView")})}))}.observes("events"),init:function(){this._super();this.updateChildViews()}});
Ember.Calendar.EventView=Ember.View.extend({templateName:"ember-calendar-event",classNames:["ember-calendar-event"],attributeBindings:["style"],style:function(){if(!this.get("event"))return"";var a=moment(this.get("event.start")).valueOf(),b=moment(this.get("event.end")).valueOf(),c=moment(a).startOf("day").valueOf()+36E5*this.get("parentView.controller.headingTimeRangeStart"),f=moment(a).startOf("day").valueOf()+36E5*this.get("parentView.controller.headingTimeRangeEnd");return"top: "+100*(a-c)/(f-
c)+"%; height: "+100*(b-a)/(f-c)+"%;"}.property("event","event.start","event.end"),nameString:function(){return this.get("event")?this.get("event.name"):""}.property("event","event.name"),timeString:function(){return this.get("event")?this.get("event.start").format(this.get("parentView.controller.eventTimeFormat"))+this.get("parentView.controller.eventTimeSeparator")+this.get("event.end").format(this.get("parentView.controller.eventTimeFormat")):""}.property("event","event.start","event.end"),locationString:function(){return this.get("event")&&
this.get("event.location")?this.get("event.location.name")||this.get("event.location.address"):""}.property("event","event.location")});


src/calendar.js
///////////////////////////////////////////////////////////////////////////////
// Namespace
///////////////////////////////////////////////////////////////////////////////
Ember.Calendar = Ember.Namespace.create();


///////////////////////////////////////////////////////////////////////////////
// Controller
///////////////////////////////////////////////////////////////////////////////
Ember.Calendar.CalendarController = Ember.ArrayController.extend({
    startOfWeek: 0
  , startOfDay: 8
  , headingDateFormat: 'ddd MMM D'
  , headingTimeFormat: 'h a'
  , headingTimeRangeStart: 0
  , headingTimeRangeEnd: 24
  , eventTimeFormat: 'h:mm a'
  , eventTimeSeparator: ' - '
  
  , headingDateViewClass: 'Ember.Calendar.HeadingDateView'
  , headingTimeViewClass: 'Ember.Calendar.HeadingTimeView'
  , dayViewClass: 'Ember.Calendar.DayView'
  , eventViewClass: 'Ember.Calendar.EventView'
  
  , week: null
  , dates: function () {
      if (!this.get('week')) {
        return [];
      }
      
      var curr = this.get('week').clone().subtract('days', 1);
      var dates = [];
      var i;
      
      for (i = 0; i < 7; i++) {
        dates.push(curr.add('days', 1).clone());
      }
      
      return dates;
    }.property('week')
  , times: function () {
      var times = [];
      var i;
      
      for (i = this.get('headingTimeRangeStart'); i < this.get('headingTimeRangeEnd'); i++) {
        times.push(1000 * 60 * 60 * i);
      }
      
      return times;
    }.property('headingTimeRangeStart', 'headingTimeRangeStart')
    
  , days: function () {
      if (!this.get('week')) {
        return [];
      }
      
      var days = [[], [], [], [], [], [], []];
      var dates = this.get('dates');
      var self = this;
      var weekStart = dates[0].clone();
      var weekEnd = dates[6].clone().endOf('day');
      
      this.get('content').forEach(function (event) {
        var start = moment(event.start).clone();
        var end = moment(event.end).clone();
        var object;
        var day;
        
        if (end < weekStart || start > weekEnd) { return; }
        
        while (end > start) {
          object = new Object();
          Object.keys(event).forEach(function (key) {
            if (key !== 'start' && key !== 'end') {
              object[key] = event[key];
            }
          });
          
          object.start = start.clone();
          object.end = start.clone().endOf('day');
          
          if (object.end > end) {
            object.end = end.clone();
          }
          
          day = object.start.clone().startOf('day').diff(self.get('week'), 'days');

          if (day >= 0 && day <= 6) {
            days[day].push(object);
          }
          
          start.add('days', 1).startOf('day');
        }
      });
      
      return days;
    }.property('content', 'dates')
    
  , actions: {
        loadPreviousWeek: function () {
          this.get('week').subtract('days', 7);
          this.notifyPropertyChange('week');
        }
      , loadNextWeek: function () {
          this.get('week').add('days', 7);
          this.notifyPropertyChange('week');
        }
    }
  
  , init: function () {
      this._super();
      this.set('week', moment(this.get('initialDate')).subtract('days', (moment(this.get('initialDate')).day() + 7 - this.get('startOfWeek')) % 7).startOf('day'));
    }
});


///////////////////////////////////////////////////////////////////////////////
// Views
///////////////////////////////////////////////////////////////////////////////
Ember.Calendar.CalendarView = Ember.View.extend({
    templateName: 'ember-calendar'
  , classNames: ['ember-calendar']
  , didInsertElement: function () {
      var container = $('#' + this.get('elementId') + ' > .ember-calendar-container');
      var body = $('#' + this.get('elementId') + ' > .ember-calendar-container > .ember-calendar-body');
      var start = (this.get('controller.startOfDay') - this.get('controller.headingTimeRangeStart')) / (this.get('controller.headingTimeRangeEnd') - this.get('controller.headingTimeRangeStart'));
      
      container.scrollTop(start * body.height());
    }
});

Ember.Calendar.HeadingDatesView = Ember.ContainerView.extend({
    classNames: ['ember-calendar-head-dates']
  , updateChildViews: function () {
      var self = this;
      self.removeAllChildren();
      self.pushObjects(self.get('dates').map(function (date) {
        return self.createChildView(Ember.get(self.get('parentView.controller.headingDateViewClass')), { date: date });
      }));
    }.observes('dates')
  , init: function () {
      this._super();
      this.updateChildViews();
    }
});

Ember.Calendar.HeadingDateView = Ember.View.extend({
    templateName: 'ember-calendar-head-date'
  , classNames: ['ember-calendar-head-date']
  , dateString: function () {
      return this.get('date').format(this.get('parentView.parentView.controller.headingDateFormat'));
    }.property('date')
});

Ember.Calendar.HeadingTimesView = Ember.ContainerView.extend({
    classNames: ['ember-calendar-head-times']
  , updateChildViews: function () {
      var self = this;
      self.removeAllChildren();
      self.pushObjects(self.get('times').map(function (time) {
        return self.createChildView(Ember.get(self.get('parentView.controller.headingTimeViewClass')), { time: time });
      }));
    }.observes('times')
  , init: function () {
      this._super();
      this.updateChildViews();
    }
});

Ember.Calendar.HeadingTimeView = Ember.View.extend({
    templateName: 'ember-calendar-head-time'
  , classNames: ['ember-calendar-head-time']
  , timeString: function () {
      return moment().startOf('day').add('milliseconds', this.get('time')).format(this.get('parentView.parentView.controller.headingTimeFormat'));
    }.property('time')
});

Ember.Calendar.DaysView = Ember.ContainerView.extend({
    classNames: ['ember-calendar-days']
  , updateChildViews: function () {
      var self = this;
      self.removeAllChildren();
      self.pushObjects(self.get('days').map(function (events) {
        return self.createChildView(Ember.get(self.get('parentView.controller.dayViewClass')), { events: events, parentView: self.get('parentView') });
      }));
    }.observes('days')
  , init: function () {
      this._super();
      this.updateChildViews();
    }
});

Ember.Calendar.DayView = Ember.ContainerView.extend({
    classNames: ['ember-calendar-day']
  , updateChildViews: function () {
      var self = this;
      self.removeAllChildren();
      self.pushObjects(self.get('events').map(function (event) {
        return self.createChildView(Ember.get(self.get('parentView.controller.eventViewClass')), { event: event, parentView: self.get('parentView') });
      }));
    }.observes('events')
  , init: function () {
      this._super();
      this.updateChildViews();
    }
});

Ember.Calendar.EventView = Ember.View.extend({
    templateName: 'ember-calendar-event'
  , classNames: ['ember-calendar-event']
  , attributeBindings: ['style']
  , style: function () {
      if (!this.get('event')) {
        return '';
      }
        
      var start = moment(this.get('event.start')).valueOf();
      var end = moment(this.get('event.end')).valueOf();
      var rangeStart = moment(start).startOf('day').valueOf() + 1000 * 60 * 60 * this.get('parentView.controller.headingTimeRangeStart');
      var rangeEnd = moment(start).startOf('day').valueOf() + 1000 * 60 * 60 * this.get('parentView.controller.headingTimeRangeEnd');
      
      return 'top: ' + 100 * (start - rangeStart) / (rangeEnd - rangeStart) + '%; height: ' + 100 * (end - start) / (rangeEnd - rangeStart) + '%;';
    }.property('event', 'event.start', 'event.end')
  , nameString: function () {
      if (!this.get('event')) {
        return '';
      }
      return this.get('event.name');
    }.property('event', 'event.name')
  , timeString: function () {
      if (!this.get('event')) {
        return '';
      }
      return this.get('event.start').format(this.get('parentView.controller.eventTimeFormat')) + this.get('parentView.controller.eventTimeSeparator') + this.get('event.end').format(this.get('parentView.controller.eventTimeFormat'));
    }.property('event', 'event.start', 'event.end')
  , locationString: function () {
      if (!this.get('event') || !this.get('event.location')) {
        return '';
      }
      return this.get('event.location.name') || this.get('event.location.address');
    }.property('event', 'event.location')
});